File: Data Structures & Algorithms/trapping-rain-water/submission-1.java
class Solution {
    public int trap(int[] height) {

        //Idea find the left most and right most value for each array element

        int n = height.length;
        int[] leftMax = getLeftMax(height, n);
        int[] rightMax = getRightMax(height, n);

        int finalSum = 0;

         //find the minimum for each element out of left and right and subtract the present height   
         for(int i = 0; i<n;i++ ){
            int h = Math.min(leftMax[i], rightMax[i]) - height[i];

            finalSum += h;
         }
         return finalSum;
    }


    int[] getLeftMax(int[] height, int n){
        int[] leftMax = new int[n];

        leftMax[0] = height[0];
        for(int i = 1;i<n;i++){
            leftMax[i] = Math.max(height[i], leftMax[i-1]);
        }
        return leftMax;
    }

    int[] getRightMax(int[] height, int n){
        int[] rightMax = new int[n];

        rightMax[n-1] = height[n-1];
        for(int i = n-2 ;i>=0;i--){
            rightMax[i] = Math.max(height[i], rightMax[i+1]);
        }
        return rightMax;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def totalHours(self, piles, mid):
        totalHours = 0
        for val in piles:
            totalHours += (val + mid-1)//mid
        return totalHours

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = max(piles)
        
        while left<right:
            mid = left + (right-left)//2
            totalHours = self.totalHours(piles, mid)

            if totalHours<=h:
                right = mid
            else:
                left = mid+1
        
        return left